### PR TITLE
 Create the e2e test namespace in each of the individual clusters in the federation.

### LINKED ISSUE #27909: Federated service shards can't be created in clusters without the specified namespace.
Federated service shards are created only in the clusters that have the specified namespace. To fix this problem we need to create the namespace on-demand while creating the federated service shards.

cc @kubernetes/sig-cluster-federation @matchstick

### PR DESCRIPTION
**Pull request by madhusudancs**
Fixes #27909 

cc @kubernetes/sig-cluster-federation @matchstick 

[Analytics]()

### CODE REVIEW

**Comment by quinton-hoole** [comment]
See comments in #27909.  We've decided not to do it this way.

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 1]
```diff
@@ -429,6 +429,21 @@ func (s *ServiceController) ensureClusterService(cachedService *cachedService, c
...
+				ns := &v1.Namespace{
+					ObjectMeta: v1.ObjectMeta{
+						Name: service.Namespace,
+					},
+				}
+				_, err := client.Core().Namespaces().Create(ns)
+				if err != nil {
+					glog.V(4).Infof("Couldn't create the namespace %s: %v", service.Namespace, err)
+				}
+				glog.V(5).Infof("Namespace %s created", service.Namespace)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/service/servicecontroller.go]: This should be in else.

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 2]
```diff
@@ -429,6 +429,21 @@ func (s *ServiceController) ensureClusterService(cachedService *cachedService, c
...
+			_, err = client.Core().Namespaces().Get(service.Namespace)
+			if errors.IsNotFound(err) {
+				ns := &v1.Namespace{
+					ObjectMeta: v1.ObjectMeta{
+						Name: service.Namespace,
+					},
+				}
+				_, err := client.Core().Namespaces().Create(ns)
+				if err != nil {
+					glog.V(4).Infof("Couldn't create the namespace %s: %v", service.Namespace, err)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/service/servicecontroller.go]: I think this is a serious error. Can remove V(4) from here.

Comment (nikhiljindal) [inline federation/pkg/federation-controller/service/servicecontroller.go]: same for the log we print when we cant create the service

**Comment by madhusudancs** [comment]
Still the same PR for the changed version. Reopening the PR. @quinton-hoole @nikhiljindal PTAL.

**Comment by quinton-hoole** [comment]
@madhusudancs We need to put this namespace creation in the e2e framework.  The e2e framework is already creating the namespace in the first cluster.  It needs to be adapted to create it in all of the clusters in the federation.  I seem to remember the @colhom wrote some federation-sepecific e2e framework bits.  This needs to go there.

**Comment by madhusudancs** [comment]
@quinton-hoole I did consider that. You can't do it there right now, unless quite a bit of code is refactored.  Individual clusters aren't registered and their clients created until BeforeEach() in the tests. So you don't have clients for the individual clusters (except `this` cluster) to make a namespace create request. It is probably a TODO for post-1.4, particularly if we want to keep the changes minimal for now.

**Comment by quinton-hoole** [comment]
@madhusudancs Cool, makes sense.  Thanks for clarifying in person.  LGTM.

**Comment by quinton-hoole** [comment]
@madhusudancs You seem to have a minor govet issue to resolve.  Feel free to re-apply the LGTM label after you've fixed that:

```
Verifying ./hack/../hack/verify-govet.sh
test/e2e/federated-service.go:131: arg err for printf verb %d of wrong type: error
exit status 1
!!! Error in ./hack/../hack/verify-govet.sh:49
  'go vet "${goflags[@]:+${goflags[@]}}" ${targets[@]}' exited with status 1
Call stack:
  1: ./hack/../hack/verify-govet.sh:49 main(...)
Exiting with status 1
FAILED   ./hack/../hack/verify-govet.sh	286s
Verifying ./hack/../hack/verify-import-boss.sh
```

**Comment by madhusudancs** [comment]
Fixed the govet issue. Adding the LGTM back.

**Comment by goltermann** [comment]
@k8s-bot node e2e test this issue: #27668

**Pull request merged**